packages/std/src/math/isqrt.rs

```diff
@@ -0,0 +1,53 @@
+use std::{cmp, ops};
+
+/// A trait for calculating the
+/// [integer square root](https://en.wikipedia.org/wiki/Integer_square_root).
+pub trait Isqrt {
+    /// The [integer square root](https://en.wikipedia.org/wiki/Integer_square_root).
+    fn isqrt(self) -> Self;
+}
+
+impl<
+        I: ops::Add<I, Output = I>
+            + ops::Div<I, Output = I>
+            + ops::Shr<u8, Output = I>
+            + cmp::PartialOrd
+            + Copy
+            + From<u8>,
+    > Isqrt for I
+{
+    /// Algorithm adapted from
+    /// [Wikipedia](https://en.wikipedia.org/wiki/Integer_square_root#Example_implementation_in_C).
+    fn isqrt(self) -> Self {
+        let mut x0 = self >> 1;
```

Comment:
Can we restrict `I` to only allow unsigned integers? The algorithm is not defined for signed integers.

Reply:
We could do that by using the `num_traits` crate and binding the [Unsigned](https://docs.rs/num-traits/0.2.14/num_traits/sign/trait.Unsigned.html) and possibly [https://docs.rs/num-traits/0.2.14/num_traits/int/trait.PrimInt.html](PrimInt) traits.

Alternatively, I could try to make `Isqrt` a marker trait with a provided method and then explicitly implement it for each uint type. What sounds better?

Reply:
Or I guess if we don't want to depend on `num_traits`, we can just create our own `Unsigned` marker trait, implement it for all the uint types, and then implement `Isqrt` generically for all `Unsigned` types. That sounds sane.

Reply:
I would avoid duplicating the implementation. Think it is nicely solved here.

But we can simply create our own marker trait for unsigned ints.

Reply:
I'm usually very careful when it comes to dependencies in the standard library. Three reasons:
- Many crates do things or pull in things that do not compile well to Wasm
- It increases the risk of pulling in float instructions even if unused. Contracts with such instructions will get rejected by the blockchain.
- Many contract developers will need to compile the dependency as part of their development process

Reply:
Ah, that explains things. Thanks. I'll be careful about dependencies then.

Redone with our own internal `Unsigned` marker.